docs/syntax-single-quotes-for-singleton-strings.md

```diff
@@ -0,0 +1,133 @@
+# Use `'` for Singleton Strings
+
+## Summary
+
+Adjust type inference to assume that string literals written using single quotes (eg `'string'`) should always be inferred to have a singleton type, and that strings written using `"double quotes"` should always be typed `string`.
+
+## Motivation
+
+It's quite tricky for us to consistently infer a singleton type in exactly all of the places where it's what the developer intended, and for us to always infer `string` where it's what the developer intended.
+
+Today, our inference strategy is to assume that a string literal should have type `string` if possible, but that type can be constrained to a singleton if necessary.
+
+This strategy is broadly successful, but fails in a few ways.
+
+### Bugs
+
+We have bugs that prevent us from deducing the proper type in certain cases.  For example: https://github.com/luau-lang/luau/issues/1483
+
+```lua
+type Form = "do-not-register" | (() -> ())
+
+local function observer(register: () -> Form) end
+
+observer(function()
+	if math.random() > 0.5 then
+		return "do-not-register" -- TypeError: Type 'string' could not be converted into '"do-not-register" | (() -> ())' Luau(1000)
+	end
+	return function() end
+end)
+```
+
+### Singleton Strings and Generics
+
+The currently implemented inference strategy doesn't always do what the developer intended in the presence of generics.  Whenever a string literal is passed to a generic, it is possible for it to be typed `string`, and so that's what Luau infers.
+
+In the following snippet using the [luaup](https://github.com/jackdotink/luaup) project, a developer clearly wanted `Kind` to be bound to a string singleton and not `string`.  They were instead forced to write a bunch of casts:
+
+```lua
+local function new_token<Kind>(kind: Kind, text: string?): cst.TokenKind<Kind>
+    return { kind = kind, text = text or kind :: any, span = span, trivia = trivias }
+end
+
+return table.freeze {
+    semicolon = new_token(';' :: any) :: cst.TokenKind<';'>,
+    equals = new_token('=' :: any) :: cst.TokenKind<'='>,
+    colon = new_token(':' :: any) :: cst.TokenKind<':'>,
+    comma = new_token(',' :: any) :: cst.TokenKind<','>,
+    dot = new_token('.' :: any) :: cst.TokenKind<'.'>,
+    endd = new_token('end' :: any) :: cst.TokenKind<'end'>,
+}
```

Comment:
For what it's worth, this could also be solved with explicit instantiation too. `new_token<<";">>(";")`. But I must say, I don't understand why `new_token(';' :: any) :: cst.TokenKind<';'>` is necessary here when `new_token(';' :: ';')` would work just as well?